### PR TITLE
net: sockets: implement getsockname function

### PR DESCRIPTION
**Pull request by tgorochowik**
From POSIX.1-2017:

The getsockname() function shall retrieve the locally-bound name of the
specified socket, store this address in the sockaddr structure pointed
to by the address argument, and store the length of this address in the
object pointed to by the address_len argument.

The address_len argument points to a socklen_t object which on input
specifies the length of the supplied sockaddr structure, and on output
specifies the length of the stored address. If the actual length of the
address is greater than the length of the supplied sockaddr structure,
the stored address shall be truncated.

If the socket has not been bound to a local name, the value stored in
the object pointed to by address is unspecified.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

### CODE REVIEW

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -1155,6 +1155,51 @@ int zsock_setsockopt(int sock, int level, int optname,
...
+			  socklen_t *addrlen)
+{
+	socklen_t newlen = 0;
+
+	/* If we don't have a connection handler, the socket is not bound */
+	if (ctx->conn_handler) {
+		SET_ERRNO(EINVAL);
+	}
+
+	if (ctx->local.family == AF_INET) {
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c]: If we check the address length validity first, then we could avoid the addr4 and addr6 stack variables.
So I am thinking something like this:
```
if (ctx->local.family == AF_INET && *addrlen != sizeof(struct in_addr)) {
	SET_ERRNO(EINVAL);
} else if (ctx->local.family == AF_INET6 && *addrlen != sizeof(struct in6_addr)) {
	SET_ERRNO(EINVAL);
}

if (IS_ENABLED(CONFIG_NET_IPV4) && ctx->local.family == AF_INET) {
	net_sin(addr)->sin_family = AF_INET;
	net_sin(addr)->sin_port = net_sin_ptr(&ctx->local)->sin_port;
	memcpy(&net_sin(addr)->sin_addr, net_sin_ptr(&ctx->local)->sin_addr,
		       sizeof(struct in_addr));

} else if (IS_ENABLED(CONFIG_NET_IPV6) && ctx->local.family == AF_INET6) {
	net_sin6(addr)->sin6_family = AF_INET6;
	net_sin6(addr)->sin6_port = net_sin6_ptr(&ctx->local)->sin6_port;
	memcpy(&net_sin(addr)->sin6_addr, net_sin6_ptr(&ctx->local)->sin6_addr,
		       sizeof(struct in6_addr));
} else {
	SET_ERRNO(EINVAL);
}

return 0;
```

What do you think?

Comment (tgorochowik) [reply subsys/net/lib/sockets/sockets.c]: But the definition says that:
> If the actual length of the address is greater than the length of the supplied sockaddr structure, the stored address shall be truncated.

That's why it is prepared in the local variables and only then copied with length set to the minimum of the size from the argument and the actual size. I don't think the local variables can be avoided - maybe by checking the it incrementally and copying field by field, but I don't think that would be any better than it is now. 

I will add the `IS_ENABLED` checks though. And I guess I should verify whether the size from the argument is not negative too.

Comment (tgorochowik) [reply subsys/net/lib/sockets/sockets.c]: BTW. this callback is often used just to check whether the socket is bound to anything, so it completely makes sense to receive a truncated address (and just ignore it) as it is the return value that matters then.

Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c]: > That's why it is prepared in the local variables and only then copied with length set to the minimum of the size from the argument and the actual size. I don't think the local variables can be avoided - maybe by checking the it incrementally and copying field by field, but I don't think that would be any better than it is now.

Ok, makes sense.

File: subsys/net/lib/sockets/sockets.c [hunk 2]
```diff
@@ -1155,6 +1155,52 @@ int zsock_setsockopt(int sock, int level, int optname,
...
+		struct sockaddr_in addr4;
+
+		addr4.sin_family = AF_INET;
+		addr4.sin_port = net_sin_ptr(&ctx->local)->sin_port;
+		memcpy(&addr4.sin_addr, net_sin_ptr(&ctx->local)->sin_addr,
+		       sizeof(struct in_addr));
+		newlen = sizeof(struct sockaddr_in);
+
+		memcpy(addr, &addr4, MIN(*addrlen, newlen));
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) &&
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c]: typo here, check should be against CONFIG_NET_IPV6

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/lib/sockets/sockets.c (hunk 2 above)
Comment (tgorochowik) [reply subsys/net/lib/sockets/sockets.c]: Right! Thanks! Fixed

**Comment by pfalcon** [comment]
Why do we need sock_getsockname_vmeth?

Commit message doesn't look right, the main thing which gets implemented is getsockname() function, not callback.

**Comment by tgorochowik** [comment]
> Commit message doesn't look right, the main thing which gets implemented is getsockname() function, not callback.

Right! Updated, thanks!

> Why do we need sock_getsockname_vmeth?

To be honest I just did what was done for other parts of `socket_op_vtable`. Do you think `getsockname` should be handled differently? Could you please elaborate?

File: include/net/socket.h [hunk 3]
```diff
@@ -370,6 +370,20 @@ int zsock_getsockopt(int sock, int level, int optname,
...
+ * @details
+ * @rststar
+ * See `POSIX.1-2017 article
+ * <http://pubs.opengroup.org/onlinepubs/9699919799/functions/getsockname.html>`__
+ * for normative description.
+ * This function is also exposed as ``getsockname()``
+ * if :option:`CONFIG_NET_SOCKETS_POSIX_NAMES` is defined.
+ * @endrststar
+ */
+int zsock_getsockname(int sock, struct sockaddr *addr, socklen_t *addrlen);
```
Comment (tbursztyka) [inline include/net/socket.h]: isn't the syscall version missing?

**Comment by pfalcon** [comment]
> To be honest I just did what was done for other parts of socket_op_vtable. Do you think getsockname should be handled differently? Could you please elaborate?

Well, I would think that local address handling is part of net_context, so should be consistent across socket family implementations (which are all currently based by net_context implementation). But yeah we need to provide an offload point anyway. But then:

1. Would be nice to emphasize that zsock_getsockname_ctx() is expected/intended to be reused across address family implementations. Perhaps, by implementing support of getsockname() for more than one AF (perhaps, add it to all existing?).
2. The idea is that socket_op_vtable is already fixed. It provides explicit vmethods for the most common operations, but the extension point to add other operations is ->ioctl(). If you look, even close() is implemented as ioctl().

**Review by pfalcon — CHANGES_REQUESTED** [review]
As described in https://github.com/zephyrproject-rtos/zephyr/pull/16100#issuecomment-491845758 , this new function almost certainly should be implemented via ->ioctl vmethod, as other operations, which aren't on critical path re: performance.

**Comment by tgorochowik** [comment]
@pfalcon thank you for your detailed comments! I have changed the implementation to use the ioctl interface. Please let me know if it is now what you had in mind

File: include/net/socket.h (hunk 3 above)
Comment (tgorochowik) [reply include/net/socket.h]: @tbursztyka right! I added the `__syscall` attribute (although it just defines it as `static inline` as far as I can see?). Is that what you had in mind?

BTW. When should the definitions using `Z_SYSCALL_HANDLER` be added? Should I add one too? Are there any guidelines for that? How do I test it after adding it?

Comment (tbursztyka) [reply include/net/socket.h]: You need to add a Z_SYSCALL_HANDLER in this case yes.
I guess the rule is: when it's an API meant to be accessible from user space, and when it access objects in kernel space, then you'll need to have this syscall handler.

**Review by pfalcon — APPROVED** [review]
>  I have changed the implementation to use the ioctl interface. 

Looks good, thanks!

File: include/net/socket.h (hunk 3 above)
Comment (tgorochowik) [reply include/net/socket.h]: Ok, thanks! I just added the syscall handler (tested it by combining the userspace demo - shared_mem and the sockets demo)

**Comment by tgorochowik** [comment]
I just realized there was a bug in the syscall handler - it did not take the possibility of address truncation and could attempt to copy too much memory, I also generally simplified the memory copying a bit. Just pushed an updated version.

Who could review the syscall code? @andrewboie ?

@jukkar Is everything alright now from your perspective?

File: subsys/net/lib/sockets/sockets.c [hunk 4]
```diff
@@ -1155,6 +1155,81 @@ int zsock_setsockopt(int sock, int level, int optname,
...
+{
+	struct sockaddr addr_cpy;
+	socklen_t addrlen_cpy, addrlen_new;
+	int ret;
+
+	addrlen_cpy = addrlen_new = *((socklen_t *)addrlen);
+
+	ret = z_impl_zsock_getsockname(sock, &addr_cpy, &addrlen_new);
+
+	Z_OOPS(z_user_to_copy((void *)addr, (void *)&addr_cpy,
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: These void* casts are superfluous.

Comment (tgorochowik) [reply subsys/net/lib/sockets/sockets.c]: The syscall arguments are integers, so without the cast you get;
```
/path/zephyr/subsys/net/lib/sockets/sockets.c:1225:24: warning: passing argument 1 of 'z_user_to_copy' makes pointer from integer without a cast [-Wint-conversion]
  Z_OOPS(z_user_to_copy(addr, (void *)&addr_cpy,
                        ^~~~
```

Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: > The syscall arguments are integers

D'oh!

Comment (tgorochowik) [reply subsys/net/lib/sockets/sockets.c]: Although the second cast is indeed unnecessary, right.

Comment (tgorochowik) [reply subsys/net/lib/sockets/sockets.c]: Removed the second cast, thanks!

**Comment by pfalcon** [comment]
> Who could review the syscall code?

Fairly speaking, you could do it yourself. You could just find an implementation of similar (by param handling) function, and compare with its syscall handler. accept() would be a good choice.

(Note that I personally specifically try to not "pick on" on things outside of the core functionality of the patch, or, as I said elsewhere, we'll have agonizingly slow development in some parts of the codebase, just like we already have - POSIX subsys is among them.)

File: subsys/net/lib/sockets/sockets.c [hunk 5]
```diff
@@ -1155,6 +1155,81 @@ int zsock_setsockopt(int sock, int level, int optname,
...
+}
+
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(zsock_getsockname, sock, addr, addrlen)
+{
+	struct sockaddr addr_cpy;
+	socklen_t addrlen_cpy, addrlen_new;
+	int ret;
+
+	addrlen_cpy = addrlen_new = *((socklen_t *)addrlen);
```
Comment (tbursztyka) [inline subsys/net/lib/sockets/sockets.c]: not super found of this way of writing, I'd pre-assign addrlen_cpy and addrlen_new directly at their declaration

**Review by tbursztyka — APPROVED** [review]
the syscall seems fine to me. Just a tiny comment, taste issue.

**Comment by tgorochowik** [comment]
@pfalcon @tbursztyka thank you for your input!

After analyzing more syscalls more thouroughly I realized that there is no need to copy anything at all. It just needs to verify if it has read/write permission on the length pointer, and write permission on the sockaddr pointer and that should be enough.
That makes the syscall significantly simpler.

**Comment by pfalcon** [comment]
> After analyzing more syscalls more thouroughly I realized that there is no need to copy anything at all. 

To support the conversation (and not call you to do more things, other folks might do that), that's not what I see. In `Z_SYSCALL_HANDLER(zsock_accept)`, I see `Z_OOPS(z_user_from_copy(&addrlen_copy, (void *)addrlen, ...))`. I.e. where you dereference a pointer from userspace directly, a reference userspace syscall handler implementation calls z_user_from_copy() to access data it points to.

I myself a userspace apprentice, and just taking an online chance to learn more of it...

**Comment by tgorochowik** [comment]
> Z_OOPS(z_user_from_copy(&addrlen_copy, (void *)addrlen, ...)). I.e. where you dereference a pointer from userspace directly, a reference userspace syscall handler implementation calls z_user_from_copy() to access data it points to.

That is right, but internally, all it does is just execute the same read/write check, and then does a regular `memcpy`.

And the very same `accept` syscall, does not operate on a copy of `addr`, it just checks whether it can write anything to it and operates on the original pointer.

So what is the reason for it? Is it because the data that is _read_ by the thread could be changed by some other thread? So we basically need a copy of everything from userspace, but anything that goes _to_ userspace doesn't need that?

I am not fully convinced, as a similar thing could happen with the address, once its set by the syscall (but before the syscall returns), some other thread could change it, and then the syscall would return an invalid value. On the other hand returning invalid values may be _less dangerous_ than operating on an invalid length.

(I am also just dumping my thoughts here, any comments explaining that would be very welcome).

**Comment by pfalcon** [comment]
> So what is the reason for it? Is it because the data that is read by the thread could be changed by some other thread? 

Yep, to save some load from @andrewboie, that's called TOCTOU stuff (https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use).

> I am not fully convinced, as a similar thing could happen with the address, once its set by the syscall (but before the syscall returns), some other thread could change it, and then the syscall would return an invalid value.

My understanding is that we care about userspace not tricking us (kernel) to operate on unchecked/purposedly hacked values. We don't care if some part of userspace plays against another part of userspace. Well, we just can't care in the same way as re: TOCTOU issues. We offer other means of [per-thread] memory protection, like "memory domains" or something.

> That is right, but internally, all it does is just ...

I personally wouldn't care about that, and just do the same as existing usages do. My attitude towards this "userspace" stuff is: it's big overhead. So, many optimality-concious people will have it disabled. And those who have it enabled, care about formal security, and then formal approach should be used: specific inventory of functions, patterns of their usage, and thoroughness of application.

**Comment by tgorochowik** [comment]
Okay, sounds good. Just pushed another version which does exactly that - copies the length, but operates on the original address pointer, so it basically a copy of what `accept` does. Thank you for your time @pfalcon

**Review by jukkar — APPROVED** [review]
LGTM

**Pull request merged**